### PR TITLE
fix(components): keep parameter values in sync with ros

### PR DESCRIPTION
**Pull request by domire8**
## Description

- #100

This PR solves the parent issue by adding a `pre_set_parameters_callback` to the component interfaces. Because the parametres in that feedback are mutable, we can basically run the whole validation in this callback, and just forward the result to the `on_set_parameters_callback`. 

It required some other changes that are rather hard to describe because of the somewhat complex call sequence in ROS but I'm happy to discuss any open points.

If you like it, I have to do the same in the modulo controllers.

## Review guidelines

Estimated Time of Review: 15 minutes

## Checklist before merging:

- [x] Confirm that the relevant changelog(s) are up-to-date in case of any user-facing changes

### CODE REVIEW

**Review by eeberhard — COMMENTED** [review]
Looking very promising in general!

File: source/modulo_components/src/ComponentInterface.cpp [hunk 1]
```diff
@@ -129,33 +134,45 @@ ComponentInterface::get_parameter(const std::string& name) const {
...
+    return;
+  }
   rcl_interfaces::msg::SetParametersResult result;
   result.successful = true;
-  for (const auto& ros_parameter : parameters) {
+  for (auto& ros_parameter : parameters) {
     try {
-      if (ros_parameter.get_name().substr(0, 27) == "qos_overrides./tf.publisher") {
+      auto parameter = parameter_map_.get_parameter(ros_parameter.get_name());
+      if (this->read_only_parameters_.at(ros_parameter.get_name())) {
```
[the remark this answers is not in the document]
Comment (domire8) [reply source/modulo_components/src/ComponentInterface.cpp:147]: From within components not, I believe. I think you could mess  with it through the ros cli but for that case it's already in a try catch block.

File: source/modulo_components/src/ComponentInterface.cpp [hunk 2]
```diff
@@ -91,32 +91,37 @@ void ComponentInterface::add_parameter(
   if (!this->node_parameters_->has_parameter(parameter->get_name())) {
     RCLCPP_DEBUG_STREAM(this->node_logging_->get_logger(), "Adding parameter '" << parameter->get_name() << "'.");
     this->parameter_map_.set_parameter(parameter);
+    this->read_only_parameters_.insert_or_assign(parameter->get_name(), false);
```
[the remark this answers is not in the document]
Comment (domire8) [reply source/modulo_components/src/ComponentInterface.cpp:94]: Kind of related to your above comment, if I don't declare here, I have an out of range exception in the pre set parameters callback.

File: source/modulo_components/src/ComponentInterface.cpp [hunk 3]
```diff
@@ -129,33 +134,45 @@ ComponentInterface::get_parameter(const std::string& name) const {
...
+            this->node_logging_->get_logger(), "Parameter '" << ros_parameter.get_name() << "' is read only.");
         continue;
       }
-      // get the associated parameter interface by name
-      auto parameter = parameter_map_.get_parameter(ros_parameter.get_name());
 
       // convert the ROS parameter into a ParameterInterface without modifying the original
       auto new_parameter = modulo_core::translators::read_parameter_const(ros_parameter, parameter);
       if (!this->validate_parameter(new_parameter)) {
+        ros_parameter = modulo_core::translators::write_parameter(new_parameter);
```
[the remark this answers is not in the document]
Comment (domire8) [reply source/modulo_components/src/ComponentInterface.cpp:156]: Good catch, added a test for that. The whole line is actually not required.

File: source/modulo_components/src/ComponentInterface.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (eeberhard) [reply source/modulo_components/src/ComponentInterface.cpp:94]: oh yeah of course, I missed the fact that `this->pre_set_parameters_callback(ros_parameters)` is called from within the block. all good then

**Review by eeberhard — APPROVED** [review]
Cool!

**Pull request merged**